metadatastore/doc.py

```diff
@@ -0,0 +1,224 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import six
+import collections
+from functools import reduce
+
+_HTML_TEMPLATE = """
+<table>
+{% for key, value in document | dictsort recursive %}
+  <tr>
+    <th> {{ key }} </th>
+    <td>
+      {% if value.items %}
+        <table>
+          {{ loop(value | dictsort) }}
+        </table>
+        {% else %}
+          {% if key == 'time' %}
+            {{ value | human_time }}
+          {% else %}
+            {{ value }}
+          {% endif %}
+        {% endif %}
+    </td>
+  </tr>
+{% endfor %}
+</table>
+"""
+
+
+class DocumentIsReadOnly(Exception):
+    pass
+
+
+class Document(dict):
+
+    def __init__(self, name, *args, **kwargs):
+        super(Document, self).__init__(*args, **kwargs)
+        super(Document, self).__setitem__('_name', name)
+        super(Document, self).__setattr__('__dict__',  self)
+
+    def __setattr__(self, key, value):
+        raise DocumentIsReadOnly()
+
+    def __setitem__(self, key, value):
+        raise DocumentIsReadOnly()
+
+    def __delattr__(self, key):
+        raise DocumentIsReadOnly()
+
+    def __delitem__(self, key):
+        raise DocumentIsReadOnly()
+
+    def update(self, *args, **kwargs):
+        raise DocumentIsReadOnly()
+
+    def pop(self, key):
+        raise DocumentIsReadOnly()
+
+    def __iter__(self):
+        return (k for k in super(Document, self).__iter__()
+                if not (isinstance(k, six.string_types) and k.startswith('_')))
+
+    def items(self):
+        return ((k, v) for k, v in super(Document, self).items()
+                if not (isinstance(k, six.string_types) and k.startswith('_')))
+
+    def values(self):
+        return (v for k, v in super(Document, self).items()
+                if not (isinstance(k, six.string_types) and k.startswith('_')))
+
+    def keys(self):
+        return (k for k in super(Document, self).keys()
+                if not (isinstance(k, six.string_types) and k.startswith('_')))
+
+    def __len__(self):
+        return len(list(self.keys()))
+
+    def _repr_html_(self):
+        import jinja2
```

Comment:
where you had the pandas import I had suggested that you might do this:

```python
try:
    pd
except NameError:
    import pandas as pd
    globals()['pd'] = pd
```

I'm sure that just got lost in the noise.  I was being lazy and trying to reference this above code snippet to avoid more typing.

Just to be explicit:

```python
try:
    jinja2
except NameError:
    import jinja2
    globals()['jinja2'] = jinja2
```

That being said, it takes 3.5 ms to import jinja2, so you might just want to ignore me here